### PR TITLE
Add support for the OpenAI API

### PR DESCRIPTION
**Pull request by levigross**
This PR enables basic support for OpenAI models (and endpoints that expose a OpenAI API compatible API interface).

I am stressing on *basic* support because we will leave tool calling to the next PR :)

### CODE REVIEW

File: model/openai/openai.go [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+		return nil, ErrModelNameRequired
+	}
+	if len(client.Options) == 0 {
+		return nil, ErrClientRequired
+	}
+	return &openAIModel{
+		client: &client,
+		name:   modelName,
+	}, nil
+}
```
[the remark this answers is not in the document]
Comment (levigross) [reply model/openai/openai.go:46]: This is wrong. `client.Options` is a slice (so `len()` will work)

File: model/openai/stream.go [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+		return nil, nil
+	}
+}
+
+// buffer is a helper that provides a strings.Builder for a given ItemID.
+// We use this to accumulate partial function call arguments as they stream in.
+func (t *streamTranslator) buffer(id string) *strings.Builder {
+	if id == "" {
+		id = "default"
+	}
```
[the remark this answers is not in the document]
Comment (levigross) [reply model/openai/stream.go:109]: The `id` parameter comes from `delta.ItemID`, which is from the OpenAI API's `ResponseFunctionCallArgumentsDeltaEvent`. It is very unlikely that OpenAI will return a blank one

File: model/openai/openai.go [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+	if err != nil {
+		return singleErrorSequence(err)
+	}
+	if stream {
+		return m.generateStream(ctx, params)
+	}
+	return m.generate(ctx, params)
+}
+
+func (m *openAIModel) generate(ctx context.Context, params responses.ResponseNewParams) iter.Seq2[*model.LLMResponse, error] {
```
Comment (cpunion) [inline model/openai/openai.go:66]: Could you please support chat completion? it's better for compatible with most providers.

**Comment by levigross** [comment]
@gemini-code-assist please rereview this code

**Comment by levigross** [comment]
@dpasiukevich What do I need to do to get this PR reviewed? 😀

**Comment by YouChenJun** [comment]
Thank you for contributing this PR to add support for the OpenAI API! This is a highly valuable feature that will greatly improve the flexibility of the project.
We noticed there are some merge conflicts now. Could you please help resolve them when you have time? This will help move the PR forward for merging.
Let us know if you need any support from our side. Thanks again for your great work! @levigross

**Comment by YouChenJun** [comment]
Thank you for contributing this PR to add support for the OpenAI API! This is a highly valuable feature that will greatly improve the flexibility of the project.
We noticed there are some merge conflicts now. Could you please help resolve them when you have time? This will help move the PR forward for merging.
Let us know if you need any support from our side. Thanks again for your great work! @levigross

**Comment by dangkaka** [comment]
hi, any updates on this?

**Comment by yafengio** [comment]
hi, any updates ?

**Comment by wjkoh** [comment]
There are currently two PRs under review and one third-party implementation:

* https://github.com/google/adk-go/pull/242
* https://github.com/google/adk-go/pull/342
* https://github.com/byebyebruce/adk-go-openai

Since the review process for both PRs seems to be lagging, could we merge a basic feature set first and improve it incrementally later? I realize that achieving full OpenAI API compatibility is a significant undertaking, and starting with a core version might be more efficient. @byebyebruce

**Comment by dpasiukevich** [comment]
Folks, sry for the lack of reply here.

We're a bit not sure on the best mechanism for long-term support of ecosystem models, let's redirect this to adk-go-community repo to unblock for now.

**Comment by dpasiukevich** [comment]
Let me setup the https://github.com/google/adk-go-community and return back here.

Let's keep and grow the stuff there with more models, tools, plugins etc.

**Comment by Kenasvarghese** [comment]
@dpasiukevich
Hi, Any update on the adk-go-community
or any other solutions to use openai models

**Comment by YouChenJun** [comment]
邮件已收到！我会尽快给您回复。

**Comment by wjkoh** [comment]
> Let me setup the https://github.com/google/adk-go-community and return back here.

Gentle reminder: Is the adk-go-community repository still coming, or has the decision changed?

**Comment by dharmjit** [comment]
Any updates on the way forward?

**Comment by marc1161** [comment]
Same here, any updates on this?

**Comment by honky-tonk** [comment]
Any updates?  guys

**Comment by bendowlingtech** [comment]
How is this still not merged? Blistering pace here.

**Comment by pmokeev** [comment]
Any updates?

**Comment by zhiyusec** [comment]
Please accelerate the progress. This issue related to model support has remained unresolved for a year. ADK-Python 2.1 now offers full model compatibility. We request urgent support for mainstream models from Google, OpenAI and Anthropic, as well as other third-party model providers.

**Comment by TunixR** [comment]
This feature is crucial for academia and industry adoption. I get that google would prefer us to use Gemini models, but that is not always possible neither optimal, so please give this feature some love. I and many others have been waiting for a serious, well-supported agent development framework for the Go language for a year at this point.

File: model/openai/openai.go [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+		}
+		llmResp := converters.Genai2LLMResponse(genaiResp)
+		attachMetadata(llmResp, resp)
+		yield(llmResp, nil)
+	}
+}
+
+func (m *openAIModel) generateStream(ctx context.Context, params responses.ResponseNewParams) iter.Seq2[*model.LLMResponse, error] {
+	return func(yield func(*model.LLMResponse, error) bool) {
+		stream := m.client.Responses.NewStreaming(ctx, params)
```
Comment (wolo-lab) [inline model/openai/openai.go:86]: Please `defer stream.Close()` to prevent leaks. In case of `!field(...)` we left stream unclosed.

File: model/openai/response.go [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+	}
+	switch resp.IncompleteDetails.Reason {
+	case "max_output_tokens":
+		return genai.FinishReasonMaxTokens
+	case "content_filter":
+		return genai.FinishReasonSafety
+	case "tool_calls_exhausted":
+		return genai.FinishReasonOther
+	case "":
+		// fallthrough
```
Comment (wolo-lab) [inline model/openai/response.go:139]: It's misleading. It doesn't work like that in go.

File: model/openai/response.go [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+	default:
+		if resp.IncompleteDetails.Reason != "" {
```
Comment (wolo-lab) [inline model/openai/response.go:141]: this condition is always true here right now

File: model/openai/response.go [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+func finishReason(resp *responses.Response) genai.FinishReason {
+	if resp == nil {
+		return genai.FinishReasonUnspecified
+	}
+	switch resp.IncompleteDetails.Reason {
+	case "max_output_tokens":
+		return genai.FinishReasonMaxTokens
+	case "content_filter":
+		return genai.FinishReasonSafety
+	case "tool_calls_exhausted":
```
Comment (wolo-lab) [inline model/openai/response.go:136]: Could you please double check if this value exist in the spec? I don't see it.

File: model/openai/openai.go [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+			yield(nil, fmt.Errorf("openai: call failed: %w", err))
+			return
+		}
+		genaiResp, err := convertResponse(resp)
+		if err != nil {
+			yield(nil, err)
+			return
+		}
+		llmResp := converters.Genai2LLMResponse(genaiResp)
+		attachMetadata(llmResp, resp)
```
Comment (wolo-lab) [inline model/openai/openai.go:79]: Shouldn't we add it in generateStream as well?

File: model/openai/doc.go [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+// Clients construct a github.com/openai/openai-go/v3 client directly and pass
+// it to NewModel:
+//
+//	ctx := context.Background()
+//	client := openai.NewClient(option.WithAPIKey(os.Getenv("OPENAI_API_KEY")))
+//	llm, err := openaimodel.NewModel(ctx, openai.ChatModelGPT4oMini, client)
+//	if err != nil {
+//		log.Fatal(err)
+//	}
+package openai
```
Comment (wolo-lab) [inline model/openai/doc.go:29]: I think we would have a collision with SDK github.com/openai-go/v2 (openai too). Maybe openaimodel? Please fix the example too.

File: model/openai/stream.go [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+		return nil, nil
+	case responseFunctionCallArgumentsDone:
+		done := evt.AsResponseFunctionCallArgumentsDone()
+		// When function call arguments are complete, we emit the full function call.
+		part, err := t.emitFunctionCall(done)
+		if err != nil {
+			return nil, err
+		}
+		return singlePartResponse(part), nil
+	case "response.failed":
```
Comment (wolo-lab) [inline model/openai/stream.go:79]: here (and in :83) we have literals, but for other cases we use constans. Let's be consistent and add them to consts.go too.

File: model/openai/openai_test.go [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+	"github.com/openai/openai-go/v3"
+	"github.com/openai/openai-go/v3/option"
+	"google.golang.org/adk/model"
+	"google.golang.org/genai"
+)
+
+func TestModel_Generate(t *testing.T) {
+	server := newLocalhostServer(t, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.URL.Path != "/v1/responses" {
+			t.Fatalf("unexpected path: %s", r.URL.Path)
```
Comment (wolo-lab) [inline model/openai/openai_test.go:36]: t.Errorf

File: model/openai/openai_test.go [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+		},
+		{
+			name:      "missing client",
+			modelName: openai.ChatModelGPT4o,
+			client:    openai.Client{},
+			wantErr:   ErrClientRequired,
+		},
+	}
+
+	ctx := context.Background()
```
Comment (wolo-lab) [inline model/openai/openai_test.go:109]: please use `t.Context()` instead

File: model/openai/openai_test.go [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+
+	for _, tt := range tests {
+		tt := tt
```
Comment (wolo-lab) [inline model/openai/openai_test.go:112]: to remove

File: model/openai/response.go [hunk 14]
```diff
@@ -0,0 +1,173 @@
...
+		},
+		CandidatesTokensDetails: []*genai.ModalityTokenCount{
+			{Modality: genai.MediaModalityText, TokenCount: int32(usage.OutputTokens)},
+		},
+		ThoughtsTokenCount: int32(usage.OutputTokensDetails.ReasoningTokens),
+	}
+}
+
+func promptFeedback(resp *responses.Response) *genai.GenerateContentResponsePromptFeedback {
+	if resp == nil || resp.IncompleteDetails.Reason == "" {
```
Comment (wolo-lab) [inline model/openai/response.go:166]: We have two reasons: `max_output_tokens` and `content_filter` and BlockedReason should be used only for one of them.

```go
	if resp == nil || resp.IncompleteDetails.Reason != "content_filter" {
		return nil
	}
```

Please add tests: ` max_output_tokens → PromptFeedback == nil and FinishReason == MaxTokens.`

File: model/openai/request.go [hunk 15]
```diff
@@ -0,0 +1,409 @@
...
+	if len(cfg.StopSequences) > 0 {
+		return ErrStopSequencesNotSupported
+	}
+	if cfg.CandidateCount > 1 {
+		return ErrMultipleCandidatesNotSupported
+	}
+	if cfg.FrequencyPenalty != nil || cfg.PresencePenalty != nil {
+		return ErrPenaltiesNotSupported
+	}
+	if cfg.ResponseLogprobs {
```
Comment (wolo-lab) [inline model/openai/request.go:303]: `ResponseLogprobs` is requested but never returned — currently a no-op

Setting `cfg.ResponseLogprobs` has no observable effect today, for two reasons.

1. Request side. We set `params.TopLogprobs`, but the Responses API only returns logprobs when `"message.output_text.logprobs"` is added to the request's Include field. Without it the API returns none:
```go
 if cfg.ResponseLogprobs {
 	if cfg.Logprobs != nil {
 		params.TopLogprobs = param.NewOpt(int64(*cfg.Logprobs))
 	} else {
 		params.TopLogprobs = param.NewOpt(int64(1))
 	}
 	// Responses returns logprobs only when explicitly included.
 	params.Include = append(params.Include, responses.ResponseIncludableMessageOutputTextLogprobs)
 }
```
2. Response side (`response.go`). Nothing reads logprobs back — `convertResponse` never populates `Candidate.LogprobsResult`, so even if returned they'd be dropped. Thread them in via `buildCandidate `:
```go
 func buildCandidate(resp *responses.Response) (*genai.Candidate, error) {
 	parts, err := convertOutputItems(resp.Output)
 	if err != nil {
 		return nil, err
 	}
 	return &genai.Candidate{
 		Content:        &genai.Content{Role: string(genai.RoleModel), Parts: parts},
 		FinishReason:   finishReason(resp),
 		LogprobsResult: convertLogprobs(resp.Output),
 	}, nil
 }
```

File: model/openai/openai.go [hunk 16]
```diff
@@ -0,0 +1,146 @@
...
+	"google.golang.org/adk/internal/llminternal/converters"
+	"google.golang.org/adk/model"
+)
+
+type openAIModel struct {
+	client *openai.Client
+	name   string
+}
+
+func NewModel(_ context.Context, modelName string, client openai.Client) (model.LLM, error) {
```
Comment (wolo-lab) [inline model/openai/openai.go:34]: Align client construction with the repo's config-struct convention

 NewModel currently takes a pre-built openai.Client and guards it with `len(client.Options) == 0 `

 Two problems:
 1. The guard is a fragile proxy — it only catches the zero-value openai.Client{}, and it relies on the undocumented invariant that openai.NewClient always prepends DefaultClientOptions (so a real client's Options is never empty). A reader can't tell why len == 0 means "not provided".
 2. It diverges from how the rest of adk-go builds models. model/gemini takes a config struct `(NewModel(ctx, modelName, *genai.ClientConfig))`; config structs are the project convention rather than passing SDK clients around.

Let's introduce a ClientConfig and builde the client internally. This deletes the fragile check and mirrors gemini:
```go
 // ClientConfig configures the OpenAI client. Mirrors model/gemini, which takes
 // *genai.ClientConfig. Empty APIKey/BaseURL fall back to the OPENAI_API_KEY /
 // OPENAI_BASE_URL env vars (handled by openai-go's default options).
 type ClientConfig struct {
 	APIKey     string
 	BaseURL    string       // for OpenAI-compatible endpoints
 	HTTPClient *http.Client // optional; e.g. for tests
 	// Options is an escape hatch for advanced openai-go request options,
 	// appended after the options derived from the fields above.
 	Options []option.RequestOption
 }
```

```go
 func NewModel(_ context.Context, modelName string, cfg *ClientConfig) (model.LLM, error) {
 	if modelName == "" {
 		return nil, ErrModelNameRequired
 	}
 	if cfg == nil {
 		cfg = &ClientConfig{}
 	}
 	var opts []option.RequestOption
 	if cfg.APIKey != "" {
 		opts = append(opts, option.WithAPIKey(cfg.APIKey))
 	}
 	if cfg.BaseURL != "" {
 		opts = append(opts, option.WithBaseURL(cfg.BaseURL))
 	}
 	if cfg.HTTPClient != nil {
 		opts = append(opts, option.WithHTTPClient(cfg.HTTPClient))
 	}
 	opts = append(opts, cfg.Options...)
 	client := openai.NewClient(opts...)
 	return &openAIModel{client: &client, name: modelName}, nil
 }
```

 - `ErrClientRequired` becomes unreachable and can be removed. Callers needing a fully custom client still can via `ClientConfig.Options`.

**Review by wolo-lab — CHANGES_REQUESTED** [review]
Thank you for your contribution. Solid PR. I've added some minor comments.

Please migrate to v2.

The test coverage is pretty low - 49%. Some examples:
* `generateSteram` - streaming path end-to-end (agregator, event errors, early-return) are not testes
* tools.go
* applyGenerationConfig - errors not covered
* converFunctioncCall, finishReason, convertUsage

**Comment by karolpiotrowicz** [comment]
Hi @levigross — thank you so much for this contribution! 🙏

Your PR laid the groundwork for native OpenAI model support in ADK Go, and I'm happy to say it has now shipped. The work landed in [#1178](https://github.com/google/adk-go/pull/1178) — which built directly on top of what you started here and credits you as a co-author — and it's available in the just-released [v2.1.0](https://github.com/google/adk-go/releases/tag/v2.1.0).

You can now use OpenAI models (and OpenAI-compatible endpoints that implement the Responses API) through the standard `model.LLM` interface via the new [`model/openaimodel`](https://github.com/google/adk-go/tree/main/model/openaimodel) package, with a runnable sample in [examples/openai/](https://github.com/google/adk-go/tree/main/examples/openai).

Since the functionality is now merged and released, I'll go ahead and close this PR as incorporated — but the foundation here is yours, and we really appreciate you kicking it off. Thanks again, and we'd love to see more contributions from you!
